gtsam_unstable/geometry/Similarity3.cpp

```diff
@@ -85,10 +86,104 @@ Point3 Similarity3::transformFrom(const Point3& p, //
   return s_ * q;
 }
 
+Pose3 Similarity3::transformFrom(const Pose3& T) const {
+  Rot3 R = R_.compose(T.rotation());
+  Point3 t = Point3(s_ * (R_ * T.translation() + t_));
+  return Pose3(R, t);
+}
+
 Point3 Similarity3::operator*(const Point3& p) const {
   return transformFrom(p);
 }
 
+// Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+Similarity3 Similarity3::AlignGivenR(const std::vector<Point3Pair>& abPointPairs, const Rot3& aRb) {
+  // calculate centroids
+  Point3 aCentroid, bCentroid;
+  std::tie(aCentroid, bCentroid) = mean(abPointPairs);
+
+  // calculate scale
+  double x = 0, y = 0;
+  Point3 aPoint, bPoint;
+  for (const Point3Pair& abPair : abPointPairs) {
+    std::tie(aPoint, bPoint) = abPair;
+    const Point3 da = aPoint - aCentroid;
+    const Point3 db = bPoint - bCentroid;
+    Vector3 Rdb = aRb * db;
+    y += da.transpose() * Rdb;
+    x += Rdb.transpose() * Rdb;
+  }
+  const double s = y / x;
+
+  // calculate translation
+  const Point3 aTb = (aCentroid - s * (aRb * bCentroid)) / s;
+  return Similarity3(aRb, aTb, s);
+}
+
+// Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+Similarity3 Similarity3::AlignGivenR(const std::vector<Point3Pair>& d_abPointPairs, const Rot3& aRb, const Point3& aCentroid, const Point3& bCentroid) {
+  // calculate scale
+  double x = 0, y = 0;
+  Point3 da, db;
+  for (const Point3Pair& d_abPair : d_abPointPairs) {
+    std::tie(da, db) = d_abPair;
+    Vector3 Rdb = aRb * db;
+    y += da.transpose() * Rdb;
+    x += Rdb.transpose() * Rdb;
+  }
+  const double s = y / x;
+
+  // calculate translation
+  const Point3 aTb = (aCentroid - s * (aRb * bCentroid)) / s;
+  return Similarity3(aRb, aTb, s);
+}
+
+// Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+Similarity3 Similarity3::Align(const std::vector<Point3Pair>& abPointPairs) {
+  const size_t n = abPointPairs.size();
+  if (n < 3) throw std::runtime_error("input should have at least 3 pairs of points");  // we need at least three pairs
+
+  // calculate centroids
+  Point3 aCentroid, bCentroid;
+  std::tie(aCentroid, bCentroid) = mean(abPointPairs);
+
+  // Add to form H matrix
+  Matrix3 H = Z_3x3;
+  Point3 aPoint, bPoint;
+  std::vector<Point3Pair> d_abPointPairs;
+  for (const Point3Pair& abPair : abPointPairs) {
+    std::tie(aPoint, bPoint) = abPair;
+    Point3 da = aPoint - aCentroid;
+    Point3 db = bPoint - bCentroid;
+    d_abPointPairs.emplace_back(da, db);
+    H += da * db.transpose();
+  }
+
+  // ClosestTo finds rotation matrix closest to H in Frobenius sense
+  Rot3 aRb = Rot3::ClosestTo(H);
+
+  return AlignGivenR(d_abPointPairs, aRb, aCentroid, bCentroid);
```

Comment:
I am now more confused and mystified than ever :-) You subtract the centroids, then pass it to AlignGivenR, which extracts the centroids again?? 

Let's reset on this: to me, there are *three* expensive operations that are done, that should probably be in their own functions:
1) subtract centroids
2) form inner products x and y
3) form outer product H
Then re-write each of the *three* align functions in terms of these three operations, and see how they can share or not share computation. I am thinking the one taking the centroids should not even exists: there is just align and alignGivenR ?

Think how you could write more functions but every function is only four (or so) lines.